iodata/formats/json.py

```diff
@@ -426,9 +475,254 @@ def _load_qcschema_input(result: dict, lit: LineIterator) -> dict:
+    input_dict = _parse_input_keys(result, lit)
+    extra_dict["input"] = input_dict["extra"]
+
+    if "molecule" not in result:
+        raise FileFormatError("{}: QCSchema Input requires 'molecule' key".format(lit.filename))
+    molecule_dict = _parse_topology_keys(result["molecule"], lit)
+    input_dict.update(molecule_dict)
+    extra_dict["molecule"] = molecule_dict["extra"]
+    input_dict["extra"] = extra_dict
+    input_dict["extra"]["schema_name"] = "qcschema_input"
+
+    return input_dict
+
+
+def _parse_input_keys(result: dict, lit: LineIterator) -> dict:
+    """Parse input keys for QCSchema input or output files.
+
+    Parameters
+    ----------
+    result
+        The JSON dict loaded from file.
+    lit
+        The line iterator holding the file data.
+
+    Returns
+    -------
+    input_dict
+        Output dictionary containing ``lot`` and ``obasis_name`` keys and corresponding values.
+        It may contain ``obasis`` and ``extra`` keys and corresponding values as well.
+
+    """
+    # QCEngineRecords input files don't actually specify a name or version
+    should_be_required_keys = {"schema_name", "schema_version"}
+    input_keys = {"molecule", "driver", "model"}
+    for key in should_be_required_keys:
+        if key not in result:
+            warn(
+                "{}: QCSchema files should have a '{}' key.".format(lit.filename, key),
+                FileFormatWarning,
+                2,
+            )
+    for key in input_keys:
+        if key not in result:
+            raise FileFormatError(
+                "{}: QCSchema `qcschema_input` file requires '{}' key".format(lit.filename, key)
+            )
+    # Store all extra keys in extra_dict and gather at end
+    input_dict = dict()
+    extra_dict = dict()
+
+    # Save schema name & version
+    extra_dict["schema_name"] = "qcschema_input"
+    try:
+        version = result["schema_version"]
+    except KeyError:
+        version = -1
+    if float(version) < 1.0:
+        warn(
+            "{}: Unknown `qcschema_input` version {}, loading may produce invalid results".format(
+                lit.filename, version
+            ),
+            FileFormatWarning,
+            2,
+        )
+    extra_dict["schema_version"] = version
+
+    # Load driver
+    extra_dict["driver"] = _parse_driver(result["driver"], lit)
+
+    # Load model & call basis helper if needed
+    model = _parse_model(result["model"], lit)
+    input_dict.update(model)
+    extra_dict["model"] = model["extra"]
+
+    # Load keywords & store
+    # Currently, only the IOData run_type attribute is specifically parsed from keywords, but this
+    # is a good space for passing additional IOData-specific keywords, given that the official spec
+    # treats this as program-specific territory.
+    if "keywords" in result:
+        keywords_dict = result["keywords"]
+        if "run_type" in keywords_dict:
+            input_dict["run_type"] = keywords_dict["run_type"]
```

Comment:
You're absolutely correct, however, not only could `run_type` be any string (or any other type), but there are absolutely no constraints on any key contained in `keywords`. Doing it this way was an attempt to define a IOData-specific way to pass this information using the QCSchema, since the point of the `keywords` is to contain program-specific keywords. 

The `driver` keyword takes one of `'energy'`, `'gradient'`, `'Hessian'`, (or `'properties'`, which exists in [QCElemental](https://github.com/MolSSI/QCElemental/blob/master/qcelemental/models/common_models.py) and in the [QCSchema online docs](https://molssi-qc-schema.readthedocs.io/en/latest/spec_components.html) but not actually in their schema on Github). The `driver` then corresponds to the value contained in the `return_result` keyword in an Output file, except for the case of `properties`, in which case `return_result` yields a redundant version of the `properties` dictionary. There is no way that I've seen that allows someone to specifically designate particular properties of interest, it seems to be all or nothing. Between that and the fact that `properties` contains lots of variability in key names depending on the type of calculation (SCF, MP2, CC, etc.), it seemed like it would be challenging to infer our `run_type` from their `driver`. 

Taking all of that into consideration, I felt that giving `run_type` a specific behaviour for IOData, and then adding documentation clarifying that expected behaviour (I'll try to have that out soon), was the best option. I'm definitely open to other options, but I didn't see any quick and easy way to convert `driver` <--> `run_type`.

Reply:
Thanks for explaining. I'd suggest the following, but I'm also open to other ideas. We could relax our documentation of `run_type` to allow also other values, not even strings. That would resolve the issue here. The documentation of the input writer should then be updated to explain how it will behave when it does not understand the value of `run_type`.

Reply:
Since we're defining IOData-specific behaviour, I think instead we should consider holding onto our current API and documenting which strings are allowable for `run_type`. There's not really any sort of other documentation for how another program might use the `keywords` section, so it would make sense to use the `run_type` value if it's one of our expected values, and otherwise only store it in `extra`, in case it's meant for another program instead.

Reply:
I hadn't thought of that possibility yet. Sounds good to me, definitely better than what I proposed.